Fixes for FWRETRACT

thinkyhead opened the pull request:
Add a new feedrate for swap recovery and clarify `FWRETRACT` documentation as suggested by @studiodyne in issue #7198.

- Add `swap_retract_recover_feedrate_mm_s` and `RETRACT_RECOVER_FEEDRATE_SWAP`.
  - Apply to `G10` and `G11`, derived from sample code in #7198.
  - Implement setting of the new value with `M208 R`.
  - Add to EEPROM and bump the EEPROM version to V40.
- Write a default EEPROM block when `FWRETRACT` is disabled.
- Allow setting swap options with `M207` / `M208` even with 1 extruder.
  - In future a swap retract may be permitted with 1 extruder.
- Don't allow flow percentage to affect retract/recover.
- Rename `MIN_RETRACT` to `MIN_AUTORETRACT`.
- If `MIN_AUTORETRACT` is set to 100 or greater, auto-retract is disabled in firmware.

---
#### Notes on auto-retract

In normal usage, the slicer will either produce normal E-only (`G0`) retract/recover moves or it will output firmware retract `G10`/`G11` codes. If the slicer is using `G10`/`G11` then it should also include `M209 S0` to turn off auto-retract. If it fails to turn off auto-retract, results will be unpredictable.

Currently, the status of `M209` is persistent and if saved to EEPROM it persists across reboot. It is expected that once you settle on a preference for firmware retraction then you'll set the machine and keep it on or off. The safer option is to leave it off, unless using legacy G-code where the retraction is poorly set and you need auto-retraction. (Perhaps it makes sense to disable auto-retract any time `G10` is received.)

Auto-retract has a `MIN_RETRACT` setting, but it has no `MAX_RETRACT` setting. So long `G0` moves for manual filament change may be seen as retract/recover moves when auto-retract is enabled. We advise turning it off, but it may not be obvious that auto-retract is involved when the feedrate is slow on E-only moves.

Also note that currently, after any auto-retraction or `G10`, the retracted extruder is marked as retracted (`G10 S1` marks it as swap-retracted). If auto-retract is enabled, the next large enough forward E move will be assumed to be a recover move (even if there were intervening smaller E forward moves).

It's tricky to avoid the side-effects of auto-retract, but we could try to make sure it's disabled more of the time. Some things that could help:
- Don't save auto-retract enabled state to EEPROM. Always have it off at reboot.
- Always turn off auto-retract in `G28`.
- Always turn off auto-retract in `G10`.

And our advice should be: **Avoid using auto-retract unless absolutely needed!**

---
We can do some things to work around the side-effects of auto-retract, but first it's most important to make sure that `G10` and `G11` work properly when auto-retract is turned off and the slicer is generating proper output. Can we confirm at this time that `G10`, `G10 S1`, and `G11` are working harmoniously together with current slicers?

---
Testers: The code in `retract()` is simplified, but still more convoluted than I suspect it needs to be, and I would like to try to bring it down to its essence. To help with that, please provide descriptions along the lines of:

- With auto-retract on, sent "G1 E-1" then "G0 E2" and XXX happened.
- Sent "G0 E2" with no retract before it and XXX happened.
- Sent "G10" followed by "G11" and XXX happened.
- Sent "G10 S1" followed by "G11" and XXX happened.
- Sent "G10 S1" followed by "G0 E2" and XXX happened.

Uncomment the debug output code in `retract()` for extra helpful output.

thinkyhead committed:
Improve retract() for G10/G11/autoretract

thinkyhead committed:
Add a Travis CI test for FWRETRACT

tcm0116 commented on Marlin/Marlin_main.cpp:
```diff
@@ -3064,52 +3045,131 @@ static void homeaxis(const AxisEnum axis) {
 
-      feedrate_mm_s = retract_recover_feedrate_mm_s;
+      // A retract multiplier has been added here to get faster swap recovery
+      feedrate_mm_s = swapping ? swap_retract_recover_feedrate_mm_s : retract_recover_feedrate_mm_s;
```
I've been thinking about this, but since you're re-working the FWRETRACT option, maybe now's a good time to bring it up.

Does the volumetric multiplier need to be applied to the retraction feed rate in a similar manner to how it is applied to the retraction length? It would seem that by not applying the volumetric multiplier, then the physical feed rate will change as a function of the filament diameter when volumetric extrusion is enabled.

thinkyhead replied:
The feedrate is always linear. When in volumetric mode, `G1 E1000` will be transparently converted to extrude whatever length corresponds to 10mm^3 (1cc) of filament, using whatever linear feedrate is set.

studiodyne replied:
My experience 👍 
FWRETRACT not a predictive science , only  ' OUR EYES' will set correctly the params
I ask you to never add arithmetic correction in the FWRETRACT
I explain ; When we use firmware retractation , we stay in front the machine while printing and we adjust the params to have a perfect ' restart' ,  i used 3 mm/high feedrate retract , and a recover at 8 mms of feedrate , but the machine slow the print , because too much time to recover
I have made a new setting , i recover at 50mms but because of the lag time of pressure , i have add 0.3mm extra restart length , and i have the exact behavior of the old params and the print is not stopped one second each time it recovers

I suggest you , to not apply any arithmetic correction to this , if the user reduce or increase the flux on the print , the risk is to reduce the pressure on restart , and if THE RETRACT RESTART is  corrupted ,  you will have a hole , or a 5mm missing outline , and so many risk to crash the print

For these reasons , it's better , to keep the initial values , the print will not be corrupted , and if the user modify the flux , he will stay in front the machine to verify if FWRETRACT need a compensation but i think not

Thks

studiodyne replied:
Please , for your users , make a comment with ' G10 S1 ' that i the command to unload 
I have searched in Google because noone knows this
If written on the ' FWRETRACT section' and in the Gcode List together , so many people will never waste one hour on Google to find it 
Thks

thinkyhead replied:
I have updated the `G10` page [on the Marlin website](http://marlinfw.org/docs/gcode/G010.html) and will soon update the RepRap wiki G-code page also.

thinkyhead replied:
> I ask you to never add arithmetic correction in the `FWRETRACT`

I see your point for the feedrate. In the case of a slower print, where the FR (feed rate) is slowed using the LCD knob, there is more time for ooze, and I agree a faster restart is preferable to promoting more ooze. Also, I agree that the concept of "feedrate" doesn't fit E-only moves. So, there will never be a feedrate multiplier applied to `FWRETRACT`.

The `flow_multiplier` is a different matter. The most common use case for `M221` flow adjustment is when the machine is badly tuned or the filament width is not exactly to specification, so the intent is to arrive at the same volume as when E steps was previously tuned. So, it makes sense to also apply the flow multiplier to retract and recover, certainly for non-swap moves.

If a flow multiplier is problematic, then the adjustment should be made with `M92 E` and flow should be set back to `M221 S100`.

studiodyne replied:
-The equation curve of  volume_multiplier and ' Restart lag pressure ' is not linear 
- The firmware retract is ' EYES SETTING' and is made to be setted in real time and each time something change during print  

Try this experience , make a cubic with no fill , only the outlines , use firmware retract and let the print run. Verify the restarted outlines when flux-=10% , and see if retract works fine as before
If the compensation works in all case of flux multiplier , add this ...
But i'm sure not ! 👍

tcm0116 commented on Marlin/Configuration_adv.h:
```diff
@@ -757,22 +757,28 @@
+ * Add G10 / G11 commands for automatic firmware-based retract / recover.
+ * Use M207 and M208 to define parameters for retract / recover.
+ *
+ * With this feature enabled, all negative E moves over the MIN_RETRACT length
```
For clarity, you might want to change this to something like "When auto retract is enabled via M209, all negative..." since the current statement implies that auto retract is enabled by default, which is not true.

thinkyhead replied:
I've been working on it and added a comment. See if it looks good now.

thinkyhead commented:
This could use some testing by @MarlinFirmware/testers-cartesian-team to make sure it doesn't have any bad side-effects with varying configurations.

FHeilmann commented on Marlin/Marlin_main.cpp:
```diff
@@ -51,23 +51,23 @@
- * G10  - Retract filament according to settings of M207
- * G11  - Retract recover filament according to settings of M208
- * G12  - Clean tool
+ * G10  - Retract filament according to settings of M20 (Requires FWRETRACT)
```
M207 instead of M20

FHeilmann replied:
I'll also try to test it with my CoreXY today or tomorrow

thinkyhead replied:
Thanks, I just found the `7` under the chair. 😉

FHeilmann commented:
Well, just tried this branch, unfortunately I cannot comment on the functionalities, since something else is wrong. After the (supposed?) unretract my printer slows to a crawl (likely a wrong speed value is set).

My config can be found here: https://github.com/FHeilmann/Marlin/tree/fw_retract_fix

this is a fresh sync from thinkyhead:bf_cleanup_config_adv

thinkyhead commented:
> After the (supposed?) unretract my printer slows to a crawl (likely a wrong speed value is set).

Sounds like something in the unretract code is the most likely cause. I'll look at it and do some testing also. It would be good to have this patch ahead of 1.1.5 / 2.0.x.

tcm0116 commented:
@thinkyhead I've been experimenting with using FWRETRACT to perform the swap retracts with the current version in Marlin, and I've noticed that it doesn't seem to be correctly restoring as my purge tower was significantly under-extruded. I tried increasing swap unretract to no avail. 

Hopefully I'll have some time tomorrow or this weekend to give this PR a try to see if the issue follows.

thinkyhead commented:
@tcm0116 I've made small tweaks, returning to more simple assumptions about which extruders need to do a swap-recover, so please test and see if your purge tower looks better. We should continue testing and patch up the code for all possible cases. Do some `G1` with auto-retract on, combine with `G10` and `G11`, do them out-of-order, etc.

I added a comment that users should turn off auto-retract with `M209 S0` before doing a manual filament change with `G0` or `G1`, because otherwise firmware retract will get invoked.

studiodyne commented:
@FHeilmann ... You just added the ' config.h ' ... or the maincpp too ? Because the swap recover speed is implemented in the maincpp , and not just in config.h

ghost commented:
@thinkyhead 
Over extrusion when restarting and a little blob on the prime tower is not a crime :-)
Don't tell us , you have the madness to make a perfect prime tower with perfect restarting after swap :-)

My settings let the swaprestart made a little under extrusion , the tower is 100% filled ,and that not affect the tower construction. Because , i use singlenozzle + volcano ....  I'm forced to make a 25x25 base tower and so many materials wasted... To swap the black color to the white perfectly .... But the white to the black PLA swap need just a 5x5 tower :-) 
Under extrusion on restart can make economy of plastic lol

ghost commented:
@FHeilmann ... Do you added ' maincpp ' , it sound like a not implemented function

ghost commented:
@thinkyhead 
in all cases , don't worry about the recover/flux_multiplier compensation , because , firmware retract is always adjustable , and it's all we need. If compensed , it's great but not absolutely needed.
The possibility of the realtime settings , is the true force of the Firmware retract , and when a user make a flux compensation , i'm sure he will verify the recover behavior to compensate if needed

No worry about it for me ,  all is adjustable !

FHeilmann commented:
@studiodyne the printer resumes normal operation if I manually send a G0 F9000, until the next retract/unretract combo happens so it is definitely a feedrate issue.

thinkyhead commented:
> Don't tell us , you have the madness to make a perfect prime tower with perfect restarting…

If it wasn't clear, I want to make sure a swap recover cannot happen by mistake, because there are cases where a "swap" has a very long retract / recover such as for a dual- or multi-bowden system going to a single nozzle, or the nozzle is attached to a **Y** feeder, where the retract and recover are much longer than usual, perhaps up to 40mm or 50mm. So, let's ensure that the logic is smart enough to always avoid doing a "swap recover" by mistake.

thinkyhead commented:
> the printer resumes normal operation if I manually send a `G0 F9000`, until the next retract/unretract combo happens so it is definitely a feedrate issue.

I looked closely, and from what I can see the `feedrate_mm_s` variable is always saved and restored by the `retract()` function. Can you see whether the `G10` or `G11` codes or auto-retract cause the feedrate to be left at the wrong value? I'm away from my machine now, or I would test it myself.

To enable debugging in the `retract()` function…
```diff
-   /* // debugging
+   //* // debugging
```

thinkyhead committed:
Don't let the flow multiplier affect retract/recover length

thinkyhead commented:
> in all cases , don't worry about the recover/flux_multiplier compensation , because , firmware retract is always adjustable , and it's all we need

Code added to always use a flow of 100 in `retract()`.

ghost commented:
@FHeilmann I have made some prints today , but always my ' z ' issue , and it was a overheated driver , i have placed a fan on the stepdrivers and all works fine NOW !!!!!
And i have forgotten to try the swapping feedratre ... shit ! I go to Holiday , next tries in two weeks 👎 
It sounds like ' autoretract activated ' , i use all days the fwretract and never had this
Autoretract must be deleted , not a good thing because of confusion when swapping

@thinkyhead : I played with my FwRetract all the day , and , we have a new issue , a little issue 
The setting of zlift in realtime force the user to apply the value only when the carriage is down , and sometimes , some retractions appears quickly , this morning , i knew this , but i have crashed a print ... To know the problem don't avoid to make it 
Please add this to avoid to crash a print , just by setting a zlift during a print

static float   current_zlit   ' beginning of the function'
REPLACE ALL zlift by  current_zlift
add the conditionnals ----
 ' IF lifted   UNLIFT with current_zlift'  
If not lifted  current zlift = zlift  to memorize the current zlifting

I knew that and i have crashed a print ,  too much gcode in the planner !!! 
I set 1.5mm of zlift , but , i have changed to 0 , but not at the good time , and nozzle stayed high , print died !

And thanks for the ' Flow_percent ' switch 👍

thinkyhead commented:
#### Notes on auto-retract

In normal usage, the slicer will either produce normal E-only (`G0`) retract/recover moves or it will output firmware retract `G10`/`G11` codes. If the slicer is using `G10`/`G11` then it should also include `M209 S0` to turn off auto-retract. If it fails to turn off auto-retract and includes E-only moves, results will be unpredictable.

Currently, the status of `M209` is persistent and if saved to EEPROM it persists across reboot. It is expected that once you settle on a preference for firmware retraction then you'll set the machine and keep it on or off. The safer option is to leave it off, unless using legacy G-code where the retraction is poorly set and you need auto-retraction. (Perhaps it makes sense to disable auto-retract any time `G10` is received.)

Auto-retract has a `MIN_RETRACT` setting, but it has no `MAX_RETRACT` setting. So long `G0` moves for manual filament change may be seen as retract/recover moves when auto-retract is enabled. We advise turning it off, but it may not be obvious that auto-retract is involved when the feedrate is slow on E-only moves.

Also note that currently, after any auto-retraction or `G10`, the retracted extruder is marked as retracted (`G10 S1` marks it as swap-retracted). If auto-retract is enabled, the next large enough forward E move will be assumed to be a recover move (even if there were intervening smaller E forward moves).

It's tricky to avoid the side-effects of auto-retract, but we could try to make sure it's disabled more of the time. Some things that could help:
- Don't save auto-retract enabled state to EEPROM. Always have it off at reboot.
- Always turn off auto-retract in `G28`.
- Always turn off auto-retract in `G10`.

And our advice should be: **Avoid using auto-retract unless absolutely needed!**

---
We can do some things to work around the side-effects of auto-retract, but first it's most important to make sure that `G10` and `G11` work properly when auto-retract is turned off and the slicer is generating proper output. Can we confirm at this time that `G10`, `G10 S1`, and `G11` are working harmoniously together with current slicers?

---
Testers: The code in `retract()` is simplified, but still more convoluted than I suspect it needs to be, and I would like to try to bring it down to its essence. To help with that, please provide descriptions along the lines of:

- With auto-retract on, sent "G1 E-1" then "G0 E2" and XXX happened.
- Sent "G0 E2" with no retract before it and XXX happened.
- Sent "G10" followed by "G11" and XXX happened.
- Sent "G10 S1" followed by "G11" and XXX happened.
- Sent "G10 S1" followed by "G0 E2" and XXX happened.

Uncomment the debug output code in `retract()` for extra helpful output.

thinkyhead committed:
Strip auto-retract for impossible M(IN|AX)_AUTORETRACT combo

thinkyhead committed:
Replace MIN_RETRACT with MIN_AUTORETRACT, MAX_AUTORETRACT

thinkyhead commented:
I've simplified the logic (and removed code that seems to work around auto-retraction issues) with some presumptions.
- First, any doubled retract or recover is simply ignored. For auto-retract, this means only the first in a series of E-only moves will count as the trigger for retract or recover.
- My second presumption is that Z hops included in old G-code will be performed in a consistent order, so for firmware retraction the Z raise/lower can be done always. However, I still added a test that the Z lower move is symmetrical to the z raise move.

thinkyhead committed:
Rename options with swap_ prefix

FHeilmann commented:
Issues I had with the previous version are gone. However, FWRetract still doesn't seem to play nice with UBL

thinkyhead commented:
@FHeilmann Is it the Z hop on retract that conflicts with UBL?

thinkyhead commented:
I'm merging this PR at this stage, since it doesn't appear to be any more broken than the previous code, and apparently does fix some things. But, we'll have to look closer to see what's conflicting with UBL specifically for the next minor release. I think that with this and a couple more small changes I can release 1.1.5. Unfortunately, with an incomplete Russian translation.

FHeilmann commented:
@thinkyhead yes, if I print with UBL and FW retract enabled, the nozzle starts moving further and further away from the print. I tested with a high hop height to verify that the hop is not skipped entirely and it seems that the hop is only partially undone.

If I disable UBL (M420 S0) or use slic3r retractions with retract-autodetect off everything works as expected.

ghost commented:
@thinkyhead 
I have read the new code and it's perfectly written and easy to understand 
HopHeight memorised, swap status protected, swapfeedrate in lcd and eeprom and gcode
zlift protected , flow protected.... 

Very Good job
"Ubl Fixes still waiting ' , i don't use it , i can't help you
I give priority to a perfect native BED , a 3mm glass with perfect holders , the simple ' linear ABL ' is enough in this case 

Thk you very much !!!

FHeilmann commented:
@thinkyhead I'll have a friend look into the russian translation today, maybe I'll be able to create a pull request so you won't have to push 1.1.5 without a proper russian translation.

ghost commented:
@thinkyhead 

Hi , i have worked this day ,and your code need corrections and improvements , i give you the idea , you make the PR

1- FwRetract, cannot unlift if a G1 Zxxx appears ' , 
      if you use SIMPLIFY3D you're DEAD lol , G10 + moves z + g11 = UNLIFTED
2- FwRetract must be locked at only one zlift and it not implemented '
      In simplify3d in the head , one G10S1 for each existing extruders to initialize them it's not serious  to see the carriage up and up and up :-)
3-) Speed of ZLIFT must be independant , because if we print PEEK at 10 mms , the zlift feedrate is reduced and take one hour to make the job
4 - No LCD manual setting for Retract_Swap_Feedrate it's a shame , you have implemented the gcode instructions , the eeprom read and write , but no LCD tweaking for it

My solutions and idea

1- FWretract Unlift if a z move between g10 and g11 ' 
    to delete           
             if (hop_amount && NEAR(hop_height, destination[Z_AXIS])) {
    and replace by  
              if (hop_amount) {
If the machine make a hop , the machine can recover , we trust the machine , we trust the gcode files , the machine make the job , if something is bad it's not the machine 

2-FwRetract must be locked at only one zlift 
      Replace by
                    if (has_zhop && !hop_amount) {
    If hop_amount not equal to zero , zlift exist and no other can be done ' easy to code '

3- zlift independant of the feedrate of the printing
           // user have to take care about max speed of zaxis without jerk and accelleration
          // that can loose steps if too high
          float temp_feedrate_mm_s=feedrate_mm_s; // backup the current feedrate 
          feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS]; // Z feedrate to max
          //
          current_position[Z_AXIS] -= retract_zlift;  // Pretend current pos is lower. Next move raises Z.
          SYNC_PLAN_POSITION_KINEMATIC();             // Set the planner to the new position
          prepare_move_to_destination(); 
          feedrate_mm_s = temp_feedrate_mm_s  ; //feedrate restoration

a temp variable created and killed , just for the function , to put the max speed for z 
The user can set the zlift by changing the ZMAX FEEDRATE of course
Have you the time to make , the gcode instruction , the eeprom , and the lcd implementation Lol , if you make this , you will have the better firmware retract system . It's not a competition i know , but this implementation make this as the best firmwareretract avalaible  👍 

4- lcd menu
            MENU_ITEM_EDIT(float3, "Ech Unret V", &swap_retract_recover_feedrate_mm_s, 1, 999);
The problem of this code , is to add  the MSG_xxxx , in all language 
You must add it in  the language unit , in all languages 
I give you the french here 

here is the code full written and tested

void retract(const bool retracting
    #if EXTRUDERS > 1
      , bool swapping = false
    #endif
  ) {

    static float hop_height,        // Remember where the Z height started
                 hop_amount = 0.0;  // Total amount lifted, for use in recover

    // Simply never allow two retracts or recovers in a row
    if (retracted[active_extruder] == retracting) return;

    #if EXTRUDERS < 2
      bool swapping = false;
    #endif
    if (!retracting) swapping = retracted_swap[active_extruder];

    /* // debugging
      SERIAL_ECHOLNPAIR("retracting ", retracting);
      SERIAL_ECHOLNPAIR("swapping ", swapping);
      SERIAL_ECHOLNPAIR("active extruder ", active_extruder);
      for (uint8_t i = 0; i < EXTRUDERS; ++i) {
        SERIAL_ECHOPAIR("retracted[", i);
        SERIAL_ECHOLNPAIR("] ", retracted[i]);
        SERIAL_ECHOPAIR("retracted_swap[", i);
        SERIAL_ECHOLNPAIR("] ", retracted_swap[i]);
      }
      SERIAL_ECHOLNPAIR("current_position[z] ", current_position[Z_AXIS]);
      SERIAL_ECHOLNPAIR("hop_amount ", hop_amount);
    //*/

    const bool has_zhop = retract_zlift > 0.01;     // Is there a hop set?

    const float old_feedrate_mm_s = feedrate_mm_s;
    const int16_t old_flow = flow_percentage[active_extruder];

    // Don't apply flow multiplication to retract/recover
    flow_percentage[active_extruder] = 100;

    // The current position will be the destination for E and Z moves
    set_destination_to_current();

    if (retracting) {
      // Remember the Z height since G-code may include its own Z-hop
      // For best results turn off Z hop if G-code already includes it
      hop_height = destination[Z_AXIS];

      // Retract by moving from a faux E position back to the current E position
      feedrate_mm_s = retract_feedrate_mm_s;
      current_position[E_AXIS] += (swapping ? swap_retract_length : retract_length) / volumetric_multiplier[active_extruder];
      sync_plan_position_e();
      prepare_move_to_destination();

      // Is a Z hop set, and has the hop not yet been done? && no double zlifting if hop_amount exist
      if (has_zhop && !hop_amount) {
        hop_amount += retract_zlift; 
        // user have to take care about max speed of zaxis without jerk and accelleration
        // that can loose steps if too high
        float temp_feedrate_mm_s=feedrate_mm_s; // backup the current feedrate 
        feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS]; // Z feedrate to max
        //
        current_position[Z_AXIS] -= retract_zlift;  // Pretend current pos is lower. Next move raises Z.
        SYNC_PLAN_POSITION_KINEMATIC();             // Set the planner to the new position
        prepare_move_to_destination(); 
        feedrate_mm_s = temp_feedrate_mm_s  ; //feedrate restoration
      }
    }
    else {
      // If a hop was done undo the hop 
                                                        //if (hop_amount && NEAR(hop_height, destination[Z_AXIS])) {
      if (hop_amount) {
        current_position[Z_AXIS] += hop_amount;     // Pretend current pos is higher. Next move lowers Z.
         // user have to take care about max speed of zaxis without jerk and accelleration
        // that can loose steps if too high
        float temp_feedrate_mm_s=feedrate_mm_s; // backup the current feedrate  
        feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS]; // Z feedrate to max
        //       
        SYNC_PLAN_POSITION_KINEMATIC();             // Set the planner to the new position
        prepare_move_to_destination();              // Lower to the old current pos
        hop_amount = 0.0;
        feedrate_mm_s = temp_feedrate_mm_s  ; // feedrate restoration
      }

      // A retract multiplier has been added here to get faster swap recovery
      feedrate_mm_s = swapping ? swap_retract_recover_feedrate_mm_s : retract_recover_feedrate_mm_s;

      const float move_e = swapping ? swap_retract_length + swap_retract_recover_length : retract_length + retract_recover_length;
      current_position[E_AXIS] -= move_e / volumetric_multiplier[active_extruder];
      sync_plan_position_e();

      prepare_move_to_destination();  // Recover E
    }

    // Restore flow and feedrate
    flow_percentage[active_extruder] = old_flow;
    feedrate_mm_s = old_feedrate_mm_s;

    // The active extruder is now retracted or recovered
    retracted[active_extruder] = retracting;

    // If swap retract/recover then update the retracted_swap flag too
    #if EXTRUDERS > 1
      if (swapping) retracted_swap[active_extruder] = retracting;
    #endif

    /* // debugging
      SERIAL_ECHOLNPAIR("retracting ", retracting);
      SERIAL_ECHOLNPAIR("swapping ", swapping);
      SERIAL_ECHOLNPAIR("active_extruder ", active_extruder);
      for (uint8_t i = 0; i < EXTRUDERS; ++i) {
        SERIAL_ECHOPAIR("retracted[", i);
        SERIAL_ECHOLNPAIR("] ", retracted[i]);
        SERIAL_ECHOPAIR("retracted_swap[", i);
        SERIAL_ECHOLNPAIR("] ", retracted_swap[i]);
      }
      SERIAL_ECHOLNPAIR("current_position[z] ", current_position[Z_AXIS]);
      SERIAL_ECHOLNPAIR("hop_amount ", hop_amount);
    //*/

  } // retract()

#endif // FWRETRACT

thinkyhead commented:
@studiodyne  I'm just making the 1.1.5 release now, then going to attempt to rebase `bugfix-2.0.x` onto the latest `bugfix-1.1.x` code. After that project is completed, I will look more closely at your detailed post and we'll get the last kinks out of `FWRETRACT`.

ghost commented:
@thinkyhead 
**Firmware Retract fixes easy to integrate to 1.1.5**  👍 

If you read my first post you can add the 1+2  with zero stress and you have an idea of the very easy code to write , because tested and fully problemless 👍 
For the 3 you need a little time to write it but if you use only ' english language ' , we will add in the future the translated unit for each language 


1. Corrupted zlift if a Z move appears between G10/G11
2. Zlifting Feedrate to max because in a slow print zlifting is slowed too
3. Zlifting Feedrate ' Gcode + LCD ' implementation '  to set the speed of feedrate indepently of the printing


Yes you can !! lolll

For the future .... To help you to improve the future release of Marlin , you will have to Write marlin to be immediatly configured , if a new release done , we add our own all in one ' UNIT.h' , with  pins + config+ bootscreen custom + adv config , in only one unit '  and we upload , the source will not need to be tweaked , If you make this , testing will be easy for ' you ' and for us . 
If new ' defined ' or anything must be added to config.h or other , create clone unit with just the new creations , and only in a stable release , we will merge all in one unit ' but always keep the rapid unit configurator '
It's not a long job , but , it will helps you when develloping and testing , it will helps all users to switch between all release in one click , you will have more users for trying

ghost commented:
# To help you the more as possible 👍 

Adv.h Line 783 :   
` #define RETRACT_ZLIFT_FEEDRATE 4        // Slow feedrate by default ' to be setted to max Z Feedrate'`

Config  store .cpp Line 1297 : 
`retract_zlift_feedrate_mm_s = RETRACT_ZLIFT_FEEDRATE ;`

Marlin . h  Line 394 : And Marlin cpp line 572
`retract_zlift_feedrate_mm_s ;       // M208 Z - G10/G11 Zlift feedrate`
Z letter is my own initiative 

Ultra lcd cpp Line 3396 :  
```
MENU_ITEM_EDIT(float3, "Ech Unret V", &swap_retract_recover_feedrate_mm_s, 1, 999);
MENU_ITEM_EDIT(float3, "Zlift V", & retract_zlift_feedrate_mm_s, 1, 25);
```
I don’t know how extract ‘ ZMAX SPEED’ on the array #define DEFAULT_MAX_FEEDRATE          {150,150,16, 150} loll , replace ‘ 25’ by zmax please

Marlin .cpp  Line 8606  gcode_m208
`if (parser.seen('Z')) retract_zlift_feedrate_mm_s = MMM_TO_MMS(parser.value_feedrate());`
Verify if good written

Config . store Line 137  
```
534  M208 Z    retract_zlift_feedrate_mm_s (float)
534 if free slot …
```

Line 534
`retract_zlift_feedrate_mm_s = 4 ;  `

 Line 545 
`EEPROM_WRITE(retract_zlift_feedrate_mm_s);`

Line 928
`EEPROM_READ(retract_zlift_feedrate_mm_s);`


And the modified code for FW ' just a few lines

 
 ```
// ONLY ONE ZLIFT allowed 👍  
 if (has_zhop && !hop_amount) {   
// ONLY ONE ZLIFT allowed 👍  

    hop_amount += retract_zlift; +

   // USER CONTROL of the zlift feedrate 👍
          float temp_feedrate_mm_s=feedrate_mm_s; // backup the current feedrate 
          feedrate_mm_s = retract_zlift_feedrate_mm_s;
    // USER CONTROL of the zlift feedrate 👍

    current_position[Z_AXIS] -= retract_zlift;  
    SYNC_PLAN_POSITION_KINEMATIC();           
    prepare_move_to_destination(); 

    // RESTORATION of the zlift feedrate 👍
            feedrate_mm_s = temp_feedrate_mm_s  ; //feedrate restoration
    // RESTORATION of the zlift feedrate 👍
  }
```

 ```
    // If a hop was done undo the hop AND NO OTHER CONDITIONS 👍
          if (hop_amount) {
    // If a hop was done undo the hop AND NO OTHER CONDITIONS 👍

    current_position[Z_AXIS] += hop_amount;     

  // USER CONTROL of the zlift feedrate 👍
          float temp_feedrate_mm_s=feedrate_mm_s; // backup the current feedrate 
          feedrate_mm_s = retract_zlift_feedrate_mm_s;
    // USER CONTROL of the zlift feedrate 👍

    SYNC_PLAN_POSITION_KINEMATIC();             // Set the planner to the new position
    prepare_move_to_destination();              // Lower to the old current pos
    hop_amount = 0.0;

    // RESTORATION of the zlift feedrate 👍
            feedrate_mm_s = temp_feedrate_mm_s  ; //feedrate restoration
    // RESTORATION of the zlift feedrate 👍
  }
```

Hope you enjoy it !
